### PR TITLE
Remove setup_requires and forced pip installs in setup.py

### LINKED ISSUE #1688: setup_requires is deprecated
### Informations

- **Qiskit Aer version**: 0.11.2
- **Python version**: 3.11.1
- **Operating system**: Linux

### What is the current behavior?

qiskit-aer's setup.py specifies `setup_requires` and expects scikit-build to install its build requirements if they are missing. scikit-build does this but provides no easy way to opt out. This behavior makes it difficult for some other environments (like conda) to avoid invoking pip and installing packages outside of its packaging system. Ideally, all of qiskit-aer's build requirements would be preinstalled to avoid scikit-build trying to install anything with pip, but `cmake` is a special case. scikit-build/setuptools try to install it with pip if it is not installed even if the real `cmake` command is already available on the `PATH`.

### Steps to reproduce the problem

Run `pip install --no-build-isolation .` in a build environment with all of qiskit-aer's build dependencies already installed except for the Python `cmake`  package (but with `cmake` installed on the `PATH`) and `PIP_NO_INDEX=1` set.

### What is the expected behavior?

scikit-build successfully builds the package.

### Suggested solutions

In [this comment](https://github.com/scikit-build/scikit-build/issues/829#issuecomment-1358318265), one of the maintainers of scikit-build recommends moving the `setup_requires` packages to `build-system.requires` in `pyproject.toml` because `setup_requires` is deprecated by scikit-build and by setuptools. In this case, pip has a better mechanism for bypassing the installation of the build requirements. It is also mentioned that the next version of scikit-build will handle the cmake dependency better (not installing it with pip if it is already on the `PATH`).

I ran into this trying to make a conda package for qiskit-aer (https://github.com/conda-forge/staged-recipes/pull/21404). It is not major issue for that case -- I am just patching `setup_requires` out of `setup.py`. It is always nicer not to carry extra patches though.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

Executing manual `pip install` commands from within `setup.py` is somewhat frowned upon from a Python packaging perspective.  By far the preferred mechanism now is to clearly state the build dependencies in a non-code format (`pyproject.toml`), and let a PEP-517-compatible builder handle virtual-environment isolation and dependency management.

We already had the dependencies declared statically, this commit just removes the residual code in `setup.py`.

### Details and comments

Fix #1688.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -3,87 +3,24 @@
...
 
-if not _DISABLE_CONAN:
-    setup_requirements.append('conan>=1.22.2')
-
-requirements = common_requirements + ['qiskit-terra>=0.21.0', 'scipy>=1.0']
+requirements = [
+    'qiskit-terra>=0.21.0',
+    'numpy>=1.16.3',
+    'scipy>=1.0',
+]
```
Comment (hhorii) [inline setup.py:23]: I just wondered that Aer uses `requirements.txt`as qiskit-terra does.

File: setup.py [hunk 2]
```diff
@@ -3,87 +3,24 @@
...
-# also build time/setup requirements and will be added to both lists
-# of requirements
-common_requirements = [
-    'numpy>=1.16.3',
-]
-
-setup_requirements = common_requirements + [
-    'scikit-build>=0.11.0',
-    'cmake!=3.17,!=3.17.0',
-    'pybind11>=2.6',
```
Comment (hhorii) [inline setup.py:76]: I guess that `pybind11` is missing after this change.

Comment (jakelishman) [reply setup.py:76]: These are build requirements only, not install requirements.  They're specified more properly in `pyproject.toml`, which is where they're largely actually getting pulled in from, except in cases where the build process deliberately manually overrides that, like the Conda channel that Will manages in the linked issue.

If they're here as well, `setuptools` decides to try and install them itself, which isn't helpful.

File: setup.py (hunk 1 above)
Comment (jakelishman) [reply setup.py:23]: I doubt there's much of a reason.  In my own projects I tend to put this information in `pyproject.toml` these days, but I wasn't trying to completely overhaul the packaging.

**Comment by jakelishman** [comment]
I need to go through and update some docs, and to fix the testing pipelines to properly use a PEP-517 builder instead of manual calls to `setup.py`.

**Review by hhorii — APPROVED** [review]

**Pull request merged**